src/item_tname.cpp

```diff
@@ -520,51 +522,104 @@ std::string category( item const &it, unsigned int /* quantity */,
+    seg_decl{ tname::segments::FOOD_STATUS, food_status },
+    seg_decl{ tname::segments::FOOD_IRRADIATED, food_irradiated },
+    seg_decl{ tname::segments::TEMPERATURE, temperature },
+    seg_decl{ tname::segments::CLOTHING_SIZE, clothing_size },
+    seg_decl{ tname::segments::FILTHY, filthy },
+    seg_decl{ tname::segments::BROKEN, segment_broken },
+    seg_decl{ tname::segments::CBM_STATUS, cbm_status },
+    seg_decl{ tname::segments::UPS, ups },
+    seg_decl{ tname::segments::WETNESS, wetness },
+    seg_decl{ tname::segments::ACTIVE, active },
+    seg_decl{ tname::segments::SEALED, sealed },
+    seg_decl{ tname::segments::FAVORITE_POST, post_asterisk },
+    seg_decl{ tname::segments::WEAPON_MODS, weapon_mods },
+    seg_decl{ tname::segments::RELIC, relic_charges },
+    seg_decl{ tname::segments::LINK, noop },
+    seg_decl{ tname::segments::VARS, vars },
+    seg_decl{ tname::segments::TECHNIQUES, noop },
+    seg_decl{ tname::segments::TAGS, tags },
 };
 
+constexpr bool all_printers_are_valid()
+{
+    for( seg_decl const &sd : segs_array ) {
+        std::size_t idx = static_cast<size_t>( sd.first );
+        if( idx >= num_segments ) {
+            return false;
+        }
+    }
+    return true;
+}
+constexpr bool segment_has_printer( tname::segments seg )
+{
+    for( seg_decl const &sai : segs_array ) {
+        if( sai.first == seg ) {
+            return true;
+        }
+    }
+    return false;
+}
+constexpr bool all_segments_have_printers( )
+{
+    for( size_t idx = 0; idx < num_segments; ++idx ) {
+        tname::segments seg = static_cast<tname::segments>( idx );
+        if( !segment_has_printer( seg ) ) {
+            return false;
+        }
+    }
+    return true;
+}
 } // namespace
 
+static_assert( all_segments_have_printers(),
+               "Every element of tname::segments (up to tname::segments::last_segment) must have "
+               "an associated printing function in segs_array" );
+
+static_assert( all_printers_are_valid(),
+               "Elements of tname::segments above tname::segments::last_segment are not printable and "
+               "must not have an associated printer in segs_array" );
+
+template<typename IS>
+struct segment_map_helper;
+template<size_t... I>
+struct segment_map_helper<std::index_sequence<I...>> {
+    static segs_map_t &get_segs_map() {
+        static segs_map_t segs_map { segs_array[I] ... };
+        return segs_map;
+    }
+};
+
 namespace tname
 {
 std::string print_segment( tname::segments segment, item const &it, unsigned int quantity,
                            segment_bitset const &segments )
 {
-    return ( *segment_map.at( segment ) )( it, quantity, segments );
+    // FIXME: replace array+function with initializer_list after Clang 12
+    static segs_map_t const &segs_map =
+        segment_map_helper<std::make_index_sequence<segs_array.size()>>::get_segs_map();
+    return ( *segs_map.at( segment ) )( it, quantity, segments );
```

Comment:
Why do we even need a map at all when we can index into the array directly now?

Reply:
You mean like [this](https://github.com/andrei8l/Cataclysm-DDA/commit/2f1029f51b1ca2b10947435cc53ad6da7ee5237a), right? The goal here was to make it easy for other contributors (who haven't seen this code already) to add/remove/reorder segments without too much trouble. After switching to array, I got a bunch of segfaults and it wasn't immediately obvious what I got wrong and I wrote this entire thing (I had to reorder the array).

I'll have to about this.

Reply:
Nevermind, I got it. I'm really dumb on this one